eNMS/services/configuration/netconf.py

```diff
@@ -0,0 +1,304 @@
+from eNMS.forms.fields import (
+    BooleanField,
+    HiddenField,
+    IntegerField,
+    SelectField,
+    StringField,
+)
+from eNMS.models.automation import ConnectionService
+
+
+class NetconfService(ConnectionService):
+    __tablename__ = "netconf_service"
+    pretty_name = "NETCONF"
+    id = db.Column(
+        Integer,
+        ForeignKey("connection_service.id"),
+        primary_key=True,
+    )
+    nc_type = db.Column(db.SmallString)
+    target = db.Column(db.SmallString)
+    default_operation = db.Column(db.SmallString)
+    test_option = db.Column(db.SmallString)
+    error_option = db.Column(db.SmallString)
+    xml_filter = db.Column(db.LargeString, default="")
+    timeout = db.Column(Integer, default=15)
+    lock = db.Column(Boolean, default=False)
+    unlock = db.Column(Boolean, default=False)
+    commit_conf = db.Column(Boolean, default=False)
+    copy_source = db.Column(db.SmallString)
+    source_url = db.Column(db.SmallString)
+    copy_destination = db.Column(db.SmallString)
+    destination_url = db.Column(db.SmallString)
+    xml_conversion = db.Column(Boolean, default=True)
+
+    __mapper_args__ = {"polymorphic_identity": "netconf_service"}
+
+    def job(self, run, payload, device=None):
+        xml_filter = run.sub(run.xml_filter, locals())
+        run.log(
+            "info",
+            "Sending NETCONF request",
+            device,
+            logger="security",
+        )
+        result = {"success": False, "result": "No NETCONF operation selected."}
+        # Connect with NCClient Manager
+        m = run.ncclient_connection(device)
+        # Lock target
+        if run.lock:
+            m.lock(target=run.target)
+        # Get full config
+        if run.nc_type == "get_config":
+            config = m.get_config(source=run.target).data_xml
+            if run.xml_conversion:
+                config = xmltodict.parse(str(config))
+            result = {
+                "success": True,
+                "result": config,
+            }
+        # Filtered get
+        if run.nc_type == "get_filtered_config":
+            get_filter = m.get(f"{xml_filter}").data_xml
+            if run.xml_conversion:
+                get_filter = xmltodict.parse(str(get_filter))
+            result = {"success": True, "result": get_filter}
+        # Push Config
+        if run.nc_type == "push_config":
+            if run.default_operation == "None":
+                default_operation = None
+            else:
+                default_operation = run.default_operation
+            if run.test_option == "None":
+                test_option = None
+            else:
+                test_option = run.test_option
+            if run.error_option == "None":
+                error_option = None
+            else:
+                error_option = run.error_option
+            push_conf = m.edit_config(
+                target=run.target,
+                config=f"{xml_filter}",
```

Comment:
this is just `xml_filter` ? or are you using f-string to convert it to a string implicitly ?? if you are `str(xml_filter)` is much less confusing